ui/lib/kv/addon/components/kv-patch-editor.js

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { tracked } from '@glimmer/tracking';
+import { A } from '@ember/array';
+import { hasWhitespace, isNonString } from 'vault/utils/validators';
+
+const WHITESPACE_WARNING =
+  "This key contains whitespace. If this is desired, you'll need to encode it with %20 in API requests.";
+
+const NON_STRING_WARNING =
+  'This value will be saved as a string. If you need to save a non-string value, please use the JSON editor.';
+
+class Kv {
+  @tracked key;
+  @tracked value;
+  @tracked state;
+  @tracked isInvalidKey;
+
+  constructor({ key, value = undefined, state = 'disabled' }) {
+    this.key = key;
+    this.value = value;
+    this.state = state;
+  }
+
+  get keyHasWarning() {
+    return hasWhitespace(this.key) ? WHITESPACE_WARNING : '';
+  }
+
+  get valueHasWarning() {
+    if (this.value === null) return '';
+    return isNonString(this.value) ? NON_STRING_WARNING : '';
+  }
+
+  reset() {
+    this.value = undefined;
+    this.state = 'disabled';
+  }
+
+  @action
+  updateValue(event) {
+    this.value = event.target.value;
+  }
+
+  @action
+  updateState(state) {
+    this.state = state;
+
+    if (state === 'deleted') {
+      this.value = null;
+    }
+  }
+}
+
+export default class KvPatchEditor extends Component {
+  @tracked patchData; // key value pairs in form
+  @tracked showSubkeys = false;
+
+  // tracked variables for new (initially empty) row of inputs
```

Comment:
nit: this reads as a run-on and at first didn't realize that these are separate sentences
```suggestion
  // tracked variables for new (initially empty) row of inputs.
```

Reply:
thanks for the suggestion!